BugFix: try and make connectExitStub(...) work as per the API

SlySven opened the pull request:
This PR should enable connectExitStub(...) to work as close to the existing published API as possible:
* `connectExitStub((integer) fromRoomID, (integer)toRoomID, (integer or string) direction)` where `direction` is the initial(s) or full, lower-case, un-hypenated ENGLISH word for one of the 12 normal exit directions - this will make a two way exit between the given direction of the `fromRoomID` room to the `toRoomID` room AND the corresponding reverse direction exit from the `toRoomID` room. The rooms need not be the same Area but they must BOTH have stub exits in the required direction.
* `connectExitStub((integer) fromRoomID, (integer) toRoomID)` - this will make a two way exit between the `fromRoomID` room to the `toRoomID` room AND the corresponding reverse direction exit from the `toRoomID` room. The rooms need not be the same Area but the `fromRoomID` must have only ONE stub exit with an opposite (reverse) direction one in the `toRoomID` one - either room can have other stub exits. Should there be more than one pair of stub exits a `nil` + error message will be produced listing the choices for the direction that can be passed to the three argument function call to make
the exits wanted.
* `connectExitStub((integer) fromRoomID, (integer or string) dirction)` - this will make a two way exit between the stub exit in the `fromRoomID` room to the *NEAREST* other room *IN THE SAME AREA* which has a stub exit in the reverse direction and which lies in the correct relative position (*except for the `in`/`out` directions where this is not relevant*). Should the direction be given as an integer in the range 1 to 12 this will be rejected (via a `nil` + error message) ~~because~~ *if* it is ambiguous ~~then~~ whether the number represents a direction or a to room ID *{depending on what stub exits the fromRoomID has and whether there is a room with the number that is between 1 an d12}*. Potentially unlike the prior code, this version properly detects whether a string or number is supplied as the direction argument.

Also:
* to allow the reporting of the direction as a number and a string in error messages the `(QString) TRoom::dirCodeToString(const int)` method has been made `static` so that it can be used in the `TLuaInterpreter` class.
* as indirectly mentioned above `(int) TLuaInterpreter::dirToNumber(lua_State*, int)` has been revised to check for a string or integer argument being examined - the prior code may not work as anticipated because it used `lua_isxxxx(...)` functions which can coerce the value they are dealing with (a number can be coerced into a string) - which messes with the logic.
* in places in the revised methods the integer constant values have been replaced with the `DIR_XXXXX` values defined in the `TRoom` class header file.
* to simplify (!) the coding the three different forms of the Lua API are implemented in three separate `(QString) TMap::connectExitStubByXxxx(...)` methods which collectively replace the original `(void) TMap::connectExitStub(...)` one. They are responsible for generating most of the Lua API error messages for this function and they indicate success by returning an empty string.

This should close #2386.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

#### Release post highlight
`connectExitStub(...)` should now work better in the two argument cases {and return meaningful error messages in all cases!}


*__Edited:__ to reflect revision to actively check what the second argument as a number might be when there are only 2 arguments.*

SlySven committed:
BugFix: try and make connectExitStub(...) work as per the API

This PR should enable connectExitStub(...) to work as close to the existing
published API as possible:
* connectExitStub((integer) fromRoomID, (integer)toRoomID,
                                             (integer or string) direction)
where direction is the initial(s) or full, lower-case, un-hypenated ENGLISH
word for one of the 12 normal exit directions - this will make a two way
exit between the given direction of the fromRoomID room to the toRoomID
room AND the corresponding reverse direction exit from the toRoomID room.
The rooms need not be the same Area but they must BOTH have stub exits in
the required direction.
* `connectExitStub((integer) fromRoomID, (integer) toRoomID)` - this will
make a two way exit between the fromRoomID room to the toRoomID room AND
the corresponding reverse direction exit from the toRoomID room. The rooms
need not be the same Area but the fromRoomID must have only ONE stub exit
with an opposite (reverse) direction one in the toRoomID one - either room
can have other stub exits. Should there be more than one pair of stub exits
a nil + error message will be produced listing the choices for the
direction that can be passed to the three argument function call to make
the exits wanted.
* `connectExitStub((integer) fromRoomID, (integer or string) dirction)`
- this will make a two way exit between the stub exit in the fromRoomID
room to the NEAREST other room IN THE SAME AREA which has a stub exit in
the reverse direction and which lies in the correct relative position
(except for the `in`/`out` directions where this is not relevant). Should
the direction be given as an integer in the range 1 to 12 this will be
rejected (via a nil + error message) because it is ambiguous then whether
the number represents a direction or a toRoomID. Potentially unlike the
prior code, this version properly detects whether a string or number is
supplied as the direction argument.

Also:
* to allow the reporting of the direction as a number and a string in error
messages the `(QString) TRoom::dirCodeToString(const int)` method has been
made `static` so that it can be used in the `TLuaInterpreter` class.
* as indirectly mentioned above
`(int) TLuaInterpreter::dirToNumber(lua_State*, int)` has been revised to
check for a string or integer argument being examined - the prior code
may not work as anticipated because it used `lua_isxxxx(...)` functions
which can coerce the value they are dealing with (a number can be coerced
into a string) - which messes with the logic.
* in places in the revised methods the integer constant values have been
replaced with the `DIR_XXXXX` values defined in the `TRoom` class header
file.
* to simplify (!) the coding the three different forms of the Lua API are
implemented in three separate `(QString) TMap::connectExitStubByXxxx(...)
methods which collectively replace the original
`(void) TMap::connectExitStub(...)` one. They are responsible for
generating most of the Lua API error messages for this function and they
indicate success by returning an empty string.

This should close #2386.
Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -2554,57 +2554,91 @@ int TLuaInterpreter::setExitStub(lua_State* L)
+            // It is a string so it is (we will assume) a direction
+            hasDirection = true;
+        } else if (lua_type(L, 2) == LUA_TNUMBER) {
+            if (lua_tonumber(L, 2) > DIR_OUT) {
```
This error is being reported because the code executed if the argument being tested is *greater* than an upper value or *less* than a lower one but the value being checked is obtained from a function call and I didn't consider it necessary to cache the function call value and test that twice. OTOH if it silences a warning I will refactor the code...

SlySven committed:
Revise: try harder to disambiguate between exit direction and room id

So that the two argument `connectExitStub(...)` works as good as it can
with the second argument being a number (between 1 and 12).

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: correct a logic/return after else error

A return was in the wrong place so a value set was never used as the
function would exit before it was accessed.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Bugfix: fix a second logic structure error

You can only skip putting in an `else` if it was actually proceeded with a
`return` - in the place I had messed up it wasn't.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
This PR needs to be updated with latest `development` code to pull in the new mandatory Lua tests - would you be able to do that?

SlySven committed:
Merge branch 'development' into BugFix_tryAndMake_connectExitStub()_workAsPerTheAPI

Needed to pull in development branch changes that include a Lua testing
framework.

Signed-off by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Spelling: fix a single spelling mistake in a TMap.cpp comment

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
@xekon able to test this?

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -2554,57 +2554,104 @@ int TLuaInterpreter::setExitStub(lua_State* L)
-        lua_pushstring(L, "connectExitStub: RoomId doesn't exist");
-        return lua_error(L);
+    if (lua_gettop(L) < 2) {
+        lua_pushfstring(L, "connectExitStub: missing argument #2 (toID as number or direction as number or string expected)");
```
```suggestion
        lua_pushfstring(L, "connectExitStub: missing argument #2 (toID as number or direction number/string expected)");
```

@Kebap your thoughts on the wording?

Kebap replied:
That's not how we do things around here anymore. We now have functions to phrase the error messages in order to unify the user experience and ease developer workflow. 
However in case of these overloaded functions, which can accept either number or string in a same position, the functions are not yet smart enough. So we need to still manually craft the error message to mimic the automatic formula.
You can test any function and leave an obligatory argument to see the error message reads like this:
`tempLineTrigger: bad argument #2 type (how many lines to match for as number expected, got 
no value!)`
So for your case we land round about here:
```suggestion
        lua_pushfstring(L, "connectExitStub: bad argument #2 type (toID as number, or direction as number/string expected, got no value!)");
```
I got no opinion on whether "or" or slash is better to inform both number and string are possible. I would not drop "as" to seperate argument name and type as that's a pretty established pattern.

SlySven replied:
TBH I hadn't looked too closely into the situation where one gets a `no value` as a/an (*non-existent*) argument type! Like Kebap says, I did have to manually construct the error message because I couldn't get the required message out of the functions because the `getVerifiedXxxx(...)` ones are only constructed for a single type - and they try to coerce the provided argument to that type - whereas this one needed to do the exact opposite (i.e. only treat it as a number if it is one and a string likewise). In this case saying that the argument is missing is simpler to read though I agree that it doesn't follow how our messages have tended to go - perhaps we might want to look at our handling of missing *required* arguments more generally?

This one can be revisited I am sure - but I think we benefit more from the immediate improvement in functionality of this PR as it is then we might lose from stalling as we argue over this detail.

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -2554,57 +2554,104 @@ int TLuaInterpreter::setExitStub(lua_State* L)
+                if (hasRoomWithNumberAsId) {
+                    if (hasExitStubWithNumberAsDirection) {
+                        return warnArgumentValue(
+                                L, __func__, QStringLiteral("%1 is too ambiguous a number to parse into a toID or a direction code as both are valid in this case").arg(lua_tonumber(L, 2)));
```
```suggestion
                                L, __func__, QStringLiteral("%1 is too ambiguous a number to parse into a toID or a direction code as both are valid in this case. If this is a direction, try providing it as a string").arg(lua_tonumber(L, 2)));
```

vadi2 reviewed (approved):
It looks good to me, thanks for fixing the bug with it. 

Do you have energy to create either C++ or Lua-based tests to ensure this complicated function keeps working in the future as intended?

SlySven commented on src/TRoom.cpp:
```diff
@@ -101,7 +101,7 @@ QString TRoom::dirCodeToDisplayName(const int dirCode) const
 }
 
-QString TRoom::dirCodeToString(const int dirCode) const
+/*static*/ QString TRoom::dirCodeToString(const int dirCode)
```
I only picked up on marking with this comment such methods in this manner recently but I think it does help as a reminder when one is doing a global search on the use of the method's name! :nerd_face:

SlySven committed:
Apply suggestions from code review

Revise: add a suggestion to the user on how to proceed on a message

Signed-off by: Stephen Lyons <slysven@virginmedia.com>

Co-authored-by: Vadim Peretokin <vperetokin@gmail.com>

SlySven commented:
> Do you have energy to create either C++ or Lua-based tests to ensure this complicated function keeps working in the future as intended?

Not really, I understand that we do now have a Lua framework available but I'm not sure I have the mental bandwidth currently to make use of it.

I'm just pleased that I have been able to fix a 2.5 year old bug....
